### PR TITLE
Add Torch example

### PR DESCRIPTION
**Pull request by JuliusKunze**
This example might be interesting for people familiar with machine learning. Again, please feel free to close this Pull Request if this is too specific.

Tested on OSX.

### CODE REVIEW

File: samples/torch/src/main/kotlin/HelloTorch.kt [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        THFloatTensor_addmm(result, 1f, added.raw, 1f, raw, matrix.raw)
+        added.dispose()
+        return FloatMatrix(result)
+    }
+
+    override fun toString() = "[${toArray().joinToString(",\n") { "[${it.joinToString { it.toString() }}]" }}]"
+}
+
+fun tensor(vararg values: Float): FloatVector {
+    val tensor = FloatVector(THFloatTensor_newWithSize1d(values.size.toLong())!!)
```
Comment (olonho) [inline samples/torch/src/main/kotlin/HelloTorch.kt]: I would suggest to rework example a bit, to make tensor disposal more automated, see for example https://github.com/JetBrains/kotlin-native/blob/master/samples/videoplayer/src/main/kotlin/Disposable.kt and its usage.

Comment (olonho) [reply samples/torch/src/main/kotlin/HelloTorch.kt]: Also using signExtend() instead of toLong() may make code more portable.

File: samples/torch/README.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+To build use `../gradlew build` or `./build.sh`.
+
+    ./downloadMNIST.sh
+
+will download and unzip the [MNIST dataset](https://en.wikipedia.org/wiki/MNIST_database) of 
+[70000 labeled handwritten digits](http://yann.lecun.com/exdb/mnist/) for training and testing a classifier.
+
+Then run 
+
+    ../gradlew run
```
Comment (olonho) [inline samples/torch/README.md]: You need to update upper level settings.gradle with smth like `include ':torch'`

File: samples/torch/README.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    
+Now
+
+    ./downloadTorch.sh
+
+will install it into `$HOME/.konan/third-party/torch` (if not yet done). 
+
+To build use `../gradlew build` or `./build.sh`.
+
+    ./downloadMNIST.sh
```
[the remark this answers is not in the document]
Comment (JuliusKunze) [reply samples/torch/README.md]: Forgot to add it. Fixed!

File: samples/torch/build.gradle [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+task downloadTorch(type: Exec) {
+    workingDir getProjectDir()
+    commandLine './downloadTorch.sh'
+}
+
+konanArtifacts {
+    interop('TorchInterop') {
+        defFile "src/main/c_interop/torch.def"
+        includeDirs "${torchHome}/include"
```
Comment (olonho) [inline samples/torch/build.gradle]: For me `includeDirs "${torchHome}/include", "${torchHome}/include/TH"` is needed for sample to compile.

File: samples/torch/src/main/kotlin/HelloTorch.kt [hunk 5]
```diff
@@ -0,0 +1,624 @@
+import kotlinx.cinterop.*
+import torch.*
+import platform.posix.*
+
+interface Disposable {
+    fun dispose()
+}
+
+abstract class FloatTensor(val raw: CPointer<THFloatTensor>) : Disposable {
```
Comment (olonho) [inline samples/torch/src/main/kotlin/HelloTorch.kt]: I'd suggest to split FloatTensor, FloatVector and disposable handling into separate files for better readability.

File: samples/torch/src/main/kotlin/HelloTorch.kt [hunk 6]
```diff
@@ -0,0 +1,624 @@
+import kotlinx.cinterop.*
+import torch.*
+import platform.posix.*
```
Comment (olonho) [inline samples/torch/src/main/kotlin/HelloTorch.kt]: Which POSIX APIs are used here?

Comment (JuliusKunze) [reply samples/torch/src/main/kotlin/HelloTorch.kt]: Methods for reading files, only loading the digit data (see MNIST class from line 517)

File: samples/torch/README.md (hunk 2 above)
Comment (JuliusKunze) [reply samples/torch/README.md]: Fixed

File: samples/torch/build.gradle (hunk 4 above)
Comment (JuliusKunze) [reply samples/torch/build.gradle]: Fixed. (Sorry for this, I always used build.sh over build.gradle)

File: samples/torch/src/main/kotlin/HelloTorch.kt (hunk 5 above)
Comment (JuliusKunze) [reply samples/torch/src/main/kotlin/HelloTorch.kt]: Done!

File: samples/torch/src/main/kotlin/Dataset.kt [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    private fun <T> read(fileName: String, action: (Int, CPointer<ByteVar>) -> T): T {
+        val filePath = directory + fileName
+        val file = fopen(fileName, "rb") ?: throw Error("Cannot read input file $filePath")
+
+        try {
+            memScoped {
+                fseek(file, 0, SEEK_END)
+                val fileSize = ftell(file)
+                println("Reading $fileSize bytes from $filePath...")
+                val buffer = allocArray<ByteVar>(fileSize)
```
Comment (olonho) [inline samples/torch/src/main/kotlin/Dataset.kt]: Probably regular ByteArray like in here would be somewhat cleaner: 
https://github.com/JetBrains/kotlinconf-spinner/blob/b9d4550d7611e732f2c54a54a3feeb4d4e1d5f7b/kotlin-native/samples/fullstack/common/src/main/kotlin/kommon/Kommon.kt#L11

File: samples/torch/README.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+## Installation
+
+To build [ATen (Torch for C)](https://github.com/pytorch/pytorch/tree/master/aten), 
+make sure you have Python 2.X and pyyaml installed:
+
+    # macOS: if you don't have pip
+    sudo easy_install pip
+    # Linux: if you don't have pip
+    apt-get -y install python-pip
```
Comment (olonho) [inline samples/torch/README.md]: Why do we need Python at all here? K/N binary only depends on C code/headers.

Comment (JuliusKunze) [reply samples/torch/README.md]: That is true, but the build process of ATen requires it, as described here: https://github.com/pytorch/pytorch/tree/master/aten

File: samples/torch/downloadTorch.sh [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    make DESTDIR=$TH_TARGET_DIRECTORY install
+
+    cd $TH_TARGET_DIRECTORY
+
+    # remove 'usr/local' prefix produced by make:
+    mv usr/local/* .
+    rm -d usr/local usr
+
+    # hack to solve "fatal error: 'generic/THNN.h' file not found" when linking, -I$<DIR>/include/THNN did not work
+    cp include/THNN/generic/THNN.h include/TH/generic/THNN.h
```
[the remark this answers is not in the document]
Comment (JuliusKunze) [reply samples/torch/downloadTorch.sh]: Interesting. If I remove this hack, compilation works through build.gradle, but not through build.sh:

Exception in thread "main" java.lang.Error: /Users/juliuskunze/.konan/third-party/torch/include/TH/THGenerateFloatType.h:13:10: fatal error: 'generic/THNN.h' file not found

Not sure why. Any idea?

[the remark this answers is not in the document]
Comment (olonho) [reply samples/torch/downloadTorch.sh]: Because gradle adds two -I after review feedback fix (via includeDirs), while build.sh still one.

File: samples/torch/src/main/kotlin/Dataset.kt (hunk 7 above)
Comment (JuliusKunze) [reply samples/torch/src/main/kotlin/Dataset.kt]: Great, that made it much better.

**Comment by JuliusKunze** [comment]
Thank you. Kotlin/Native is also neat btw! (:

I guess you are right, it could be a project on its own. For now, I don't have the intention to create a fully-fledged Torch client for Kotlin/Native though.

File: samples/torch/downloadTorch.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (olonho) [reply samples/torch/downloadTorch.sh]: Use smth like -copt "-I$TH_TARGET_DIRECTORY/include -I$TH_TARGET_DIRECTORY/include/TH" with quotes, it shall work.

[the remark this answers is not in the document]
Comment (JuliusKunze) [reply samples/torch/downloadTorch.sh]: Turns out linking with konanc already works, what does not work is stub generation.
``cinterop -def $DIR/src/main/c_interop/torch.def \
    -copt "-I$TH_TARGET_DIRECTORY/include -I$TH_TARGET_DIRECTORY/include/TH" \
    -labels $TARGET -o $DIR/build/c_interop/torch || exit 1``

fails with 
`Exception in thread "main" java.lang.Error: /var/folders/2m/_whs043j3wb869nhv7wk6dv00000gn/T/tmp4064000223995487764.c:1:10: fatal error: 'TH/TH.h' file not found`
